core/src/main/scala/filodb.core/memstore/TimeSeriesShard.scala

```diff
@@ -790,7 +817,42 @@ class TimeSeriesShard(val ref: DatasetRef,
     shardStats.purgedPartitions.increment(numDeleted)
   }
 
-  def createFlushTask(flushGroup: FlushGroup): Task[Response] = {
+  /**
+    * Creates zero or more flush tasks (up to the number of flush groups) based on examination
+    * of the record container's ingestion time. This should be called before ingesting the container.
+    *
+    * Note that the tasks returned by this method aren't executed yet. The caller decides how
+    * to run the tasks, and by which threads.
+    */
+  def createFlushTasks(container: RecordContainer): Seq[Task[Response]] = {
+    val tasks = new ArrayBuffer[Task[Response]]()
+
+    var oldTimestamp = lastIngestionTime
+    val ingestionTime = Math.max(oldTimestamp, container.timestamp) // monotonic clock
+    var newTimestamp = ingestionTime
+
+    if (newTimestamp > oldTimestamp && oldTimestamp != Long.MinValue) {
+      for (group <- 0 until numGroups optimized) {
+        if (oldTimestamp / flushBoundaryMillis != newTimestamp / flushBoundaryMillis) {
+          // Flush out the group before ingesting records for a new hour (by group offset).
+          tasks += createFlushTask(prepareFlushGroup(group))
+        }
+        oldTimestamp -= flushOffsetMillis
+        newTimestamp -= flushOffsetMillis
+      }
+    }
+
+    // Only update stuff if no exception was thrown.
+
+    if (ingestionTime != lastIngestionTime) {
+        lastIngestionTime = ingestionTime
```

Comment:
Can we update gauge much less frequently? No need to do for each for each kafka message. Instead once during flushing of group should be sufficient. Multiple updates of gauge within same reporting period is almost useless.

Reply:
When a gauge is updated, is a network message sent, or does it just update a number? I assumed that just a number is updated, which is cheap. Anything I do here is to reduce updating a number isn't likely to have much benefit.

Reply:
I actually didnt ask for additional check per sample.
If we simply moved the setting of gauge into the createFlushTask method or nearby, it is sufficient.

Reply:
The problem is that if I move to createFlushTask, I don't have the two timestamps to compare anymore.